// rsocket-core/dist/RequestResponseStream.js
"use strict";
/*
 * Copyright 2021-2022 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestResponseResponderStream = exports.RequestResponseRequesterStream = void 0;
var Errors_1 = require("./Errors");
var Fragmenter_1 = require("./Fragmenter");
var Frames_1 = require("./Frames");
var Reassembler = __importStar(require("./Reassembler"));
var RequestResponseRequesterStream = /** @class */ (function () {
    function RequestResponseRequesterStream(payload, receiver, fragmentSize, leaseManager) {
        this.payload = payload;
        this.receiver = receiver;
        this.fragmentSize = fragmentSize;
        this.leaseManager = leaseManager;
        this.streamType = Frames_1.FrameTypes.REQUEST_RESPONSE;
    }
    RequestResponseRequesterStream.prototype.handleReady = function (streamId, stream) {
        var e_1, _a;
        if (this.done) {
            return false;
        }
        this.streamId = streamId;
        this.stream = stream;
        stream.connect(this);
        if ((0, Fragmenter_1.isFragmentable)(this.payload, this.fragmentSize, Frames_1.FrameTypes.REQUEST_RESPONSE)) {
            try {
                for (var _b = __values((0, Fragmenter_1.fragment)(streamId, this.payload, this.fragmentSize, Frames_1.FrameTypes.REQUEST_RESPONSE)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var frame = _c.value;
                    this.stream.send(frame);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        else {
            this.stream.send({
                type: Frames_1.FrameTypes.REQUEST_RESPONSE,
                data: this.payload.data,
                metadata: this.payload.metadata,
                flags: this.payload.metadata ? Frames_1.Flags.METADATA : 0,
                streamId: streamId,
            });
        }
        if (this.hasExtension) {
            this.stream.send({
                type: Frames_1.FrameTypes.EXT,
                streamId: streamId,
                extendedContent: this.extendedContent,
                extendedType: this.extendedType,
                flags: this.flags,
            });
        }
        return true;
    };
    RequestResponseRequesterStream.prototype.handleReject = function (error) {
        if (this.done) {
            return;
        }
        this.done = true;
        this.receiver.onError(error);
    };
    RequestResponseRequesterStream.prototype.handle = function (frame) {
        var errorMessage;
        var frameType = frame.type;
        switch (frameType) {
            case Frames_1.FrameTypes.PAYLOAD: {
                var hasComplete = Frames_1.Flags.hasComplete(frame.flags);
                var hasPayload = Frames_1.Flags.hasNext(frame.flags);
                if (hasComplete || !Frames_1.Flags.hasFollows(frame.flags)) {
                    this.done = true;
                    this.stream.disconnect(this);
                    if (!hasPayload) {
                        // TODO: add validation no frame in reassembly
                        this.receiver.onComplete();
                        return;
                    }
                    var payload = this.hasFragments
                        ? Reassembler.reassemble(this, frame.data, frame.metadata)
                        : {
                            data: frame.data,
                            metadata: frame.metadata,
                        };
                    this.receiver.onNext(payload, true);
                    return;
                }
                if (!Reassembler.add(this, frame.data, frame.metadata)) {
                    errorMessage = "Unexpected fragment size";
                    break;
                }
                return;
            }
            case Frames_1.FrameTypes.ERROR: {
                this.done = true;
                this.stream.disconnect(this);
                Reassembler.cancel(this);
                this.receiver.onError(new Errors_1.RSocketError(frame.code, frame.message));
                return;
            }
            case Frames_1.FrameTypes.EXT: {
                if (this.hasFragments) {
                    errorMessage = "Unexpected frame type [".concat(frameType, "] during reassembly");
                    break;
                }
                this.receiver.onExtension(frame.extendedType, frame.extendedContent, Frames_1.Flags.hasIgnore(frame.flags));
                return;
            }
            default: {
                errorMessage = "Unexpected frame type [".concat(frameType, "]");
            }
        }
        this.close(new Errors_1.RSocketError(Errors_1.ErrorCodes.CANCELED, errorMessage));
        this.stream.send({
            type: Frames_1.FrameTypes.CANCEL,
            streamId: this.streamId,
            flags: Frames_1.Flags.NONE,
        });
        this.stream.disconnect(this);
        // TODO: throw an exception if strict frame handling mode
    };
    RequestResponseRequesterStream.prototype.cancel = function () {
        var _a;
        if (this.done) {
            return;
        }
        this.done = true;
        if (!this.streamId) {
            (_a = this.leaseManager) === null || _a === void 0 ? void 0 : _a.cancelRequest(this);
            return;
        }
        this.stream.send({
            type: Frames_1.FrameTypes.CANCEL,
            flags: Frames_1.Flags.NONE,
            streamId: this.streamId,
        });
        this.stream.disconnect(this);
        Reassembler.cancel(this);
    };
    RequestResponseRequesterStream.prototype.onExtension = function (extendedType, content, canBeIgnored) {
        if (this.done) {
            return;
        }
        if (!this.streamId) {
            this.hasExtension = true;
            this.extendedType = extendedType;
            this.extendedContent = content;
            this.flags = canBeIgnored ? Frames_1.Flags.IGNORE : Frames_1.Flags.NONE;
            return;
        }
        this.stream.send({
            streamId: this.streamId,
            type: Frames_1.FrameTypes.EXT,
            extendedType: extendedType,
            extendedContent: content,
            flags: canBeIgnored ? Frames_1.Flags.IGNORE : Frames_1.Flags.NONE,
        });
    };
    RequestResponseRequesterStream.prototype.close = function (error) {
        if (this.done) {
            return;
        }
        this.done = true;
        Reassembler.cancel(this);
        if (error) {
            this.receiver.onError(error);
        }
        else {
            this.receiver.onComplete();
        }
    };
    return RequestResponseRequesterStream;
}());
exports.RequestResponseRequesterStream = RequestResponseRequesterStream;
var RequestResponseResponderStream = /** @class */ (function () {
    function RequestResponseResponderStream(streamId, stream, fragmentSize, handler, frame) {
        this.streamId = streamId;
        this.stream = stream;
        this.fragmentSize = fragmentSize;
        this.handler = handler;
        this.streamType = Frames_1.FrameTypes.REQUEST_RESPONSE;
        stream.connect(this);
        if (Frames_1.Flags.hasFollows(frame.flags)) {
            Reassembler.add(this, frame.data, frame.metadata);
            return;
        }
        var payload = {
            data: frame.data,
            metadata: frame.metadata,
        };
        try {
            this.receiver = handler(payload, this);
        }
        catch (error) {
            this.onError(error);
        }
    }
    RequestResponseResponderStream.prototype.handle = function (frame) {
        var _a;
        var errorMessage;
        if (!this.receiver || this.hasFragments) {
            if (frame.type === Frames_1.FrameTypes.PAYLOAD) {
                if (Frames_1.Flags.hasFollows(frame.flags)) {
                    if (Reassembler.add(this, frame.data, frame.metadata)) {
                        return;
                    }
                    errorMessage = "Unexpected fragment size";
                }
                else {
                    var payload = Reassembler.reassemble(this, frame.data, frame.metadata);
                    try {
                        this.receiver = this.handler(payload, this);
                    }
                    catch (error) {
                        this.onError(error);
                    }
                    return;
                }
            }
            else {
                errorMessage = "Unexpected frame type [".concat(frame.type, "] during reassembly");
            }
        }
        else if (frame.type === Frames_1.FrameTypes.EXT) {
            this.receiver.onExtension(frame.extendedType, frame.extendedContent, Frames_1.Flags.hasIgnore(frame.flags));
            return;
        }
        else {
            errorMessage = "Unexpected frame type [".concat(frame.type, "]");
        }
        this.done = true;
        (_a = this.receiver) === null || _a === void 0 ? void 0 : _a.cancel();
        if (frame.type !== Frames_1.FrameTypes.CANCEL && frame.type !== Frames_1.FrameTypes.ERROR) {
            this.stream.send({
                type: Frames_1.FrameTypes.ERROR,
                flags: Frames_1.Flags.NONE,
                code: Errors_1.ErrorCodes.CANCELED,
                message: errorMessage,
                streamId: this.streamId,
            });
        }
        this.stream.disconnect(this);
        Reassembler.cancel(this);
        // TODO: throws if strict
    };
    RequestResponseResponderStream.prototype.onError = function (error) {
        if (this.done) {
            console.warn("Trying to error for the second time. ".concat(error ? "Dropping error [".concat(error, "].") : ""));
            return;
        }
        this.done = true;
        this.stream.send({
            type: Frames_1.FrameTypes.ERROR,
            flags: Frames_1.Flags.NONE,
            code: error instanceof Errors_1.RSocketError
                ? error.code
                : Errors_1.ErrorCodes.APPLICATION_ERROR,
            message: error.message,
            streamId: this.streamId,
        });
        this.stream.disconnect(this);
    };
    RequestResponseResponderStream.prototype.onNext = function (payload, isCompletion) {
        var e_2, _a;
        if (this.done) {
            return;
        }
        this.done = true;
        // TODO: add payload size validation
        if ((0, Fragmenter_1.isFragmentable)(payload, this.fragmentSize, Frames_1.FrameTypes.PAYLOAD)) {
            try {
                for (var _b = __values((0, Fragmenter_1.fragment)(this.streamId, payload, this.fragmentSize, Frames_1.FrameTypes.PAYLOAD, true)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var frame = _c.value;
                    this.stream.send(frame);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        else {
            this.stream.send({
                type: Frames_1.FrameTypes.PAYLOAD,
                flags: Frames_1.Flags.NEXT | Frames_1.Flags.COMPLETE | (payload.metadata ? Frames_1.Flags.METADATA : 0),
                data: payload.data,
                metadata: payload.metadata,
                streamId: this.streamId,
            });
        }
        this.stream.disconnect(this);
    };
    RequestResponseResponderStream.prototype.onComplete = function () {
        if (this.done) {
            return;
        }
        this.done = true;
        this.stream.send({
            type: Frames_1.FrameTypes.PAYLOAD,
            flags: Frames_1.Flags.COMPLETE,
            streamId: this.streamId,
            data: null,
            metadata: null,
        });
        this.stream.disconnect(this);
    };
    RequestResponseResponderStream.prototype.onExtension = function (extendedType, content, canBeIgnored) {
        if (this.done) {
            return;
        }
        this.stream.send({
            type: Frames_1.FrameTypes.EXT,
            streamId: this.streamId,
            flags: canBeIgnored ? Frames_1.Flags.IGNORE : Frames_1.Flags.NONE,
            extendedType: extendedType,
            extendedContent: content,
        });
    };
    RequestResponseResponderStream.prototype.close = function (error) {
        var _a;
        if (this.done) {
            console.warn("Trying to close for the second time. ".concat(error ? "Dropping error [".concat(error, "].") : ""));
            return;
        }
        Reassembler.cancel(this);
        (_a = this.receiver) === null || _a === void 0 ? void 0 : _a.cancel();
    };
    return RequestResponseResponderStream;
}());
exports.RequestResponseResponderStream = RequestResponseResponderStream;
